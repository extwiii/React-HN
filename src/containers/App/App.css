@import url('https://fonts.googleapis.com/css?family=Lato');

body {
  font-family: 'Lato', sans-serif;
  background-color: #fafbfc;
}

header {
  display: flex;
  padding-left: 1em;
  align-items: center;
  height: 50px;
  background: linear-gradient(135deg, #ff6600, red);
  box-shadow: 10px 10px 25px 0px rgba(46, 61, 73, 0.2);
  color: white;
}

header img {
  border: 1px solid white;
}

header h1 {
  font-size: 1em;
  padding-left: 1em;
}

header a {
  color: white;
  text-decoration: none;
  cursor: pointer;
}

.App {
  margin: 1em;
  display: grid;
  grid-template-columns: repeat(1, 1fr);
}

.story_item.open {
  grid-row: span 3;
}

.spinner {
  margin: 1em;
  display: flex;
  margin-top: 50%;
  justify-content: center;
}

.pagination {
  padding: 0;
  margin: 2em;
  display: flex;
  justify-content: center;
}

.pagination_item {
  padding: 0.75em;
  display: inline;
}

.pagination_link {
  font-size: 1.25em;
  text-decoration: none;
  color: #ff6600;
}

.pagination_link--active {
  padding: 3px 10px;
  border-radius: 3px;
  background: linear-gradient(135deg, #ff6600, red);
  color: white;
  font-weight: bold;
}

@media (min-width: 576px) {
  .App {
    grid-template-columns: repeat(2, 1fr);
  }
  .story_item.open {
    animation: opacity 0.5s ease-in-out;
    grid-column: span 2;
    grid-row: span 3;
  }
}

@media (min-width: 768px) {
  .spinner {
    margin-top: 20%;
  }
  .App {
    grid-template-columns: repeat(3, 1fr);
  }
}

@media (min-width: 1200px) {
  .App {
    grid-template-columns: repeat(5, 1fr);
  }
}

@keyframes opacity {
  0% {
    opacity: 0.8;
  }
  100% {
    opacity: 1;
  }
}
