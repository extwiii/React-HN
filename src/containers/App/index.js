// @flow
import React, { Component } from 'react';
import axios from 'axios';
import Pagination from 'react-js-pagination';

import './App.css';

import StoryItem from '../../components/StoryItem';

const HN_TOP_STORIED_URL = 'https://hacker-news.firebaseio.com/v0/topstories.json';
const HN_STORY_URL = 'https://hacker-news.firebaseio.com/v0/item';
const itemArray = [];
let textArray = [];

type KidsItems = {
  by: string,
  id: number,
  kids: ?Array<number>,
  parent: number,
  text: string,
  time: number,
  type: string,
};

type TopStories = {
  by: string,
  descendants: number,
  id: number,
  kids: ?Array<number>,
  kidsItems: ?Array<KidsItems>,
  score: number,
  time: number,
  title: string,
  type: string,
  url: string,
};
type State = {
  activePage: number,
  isCompleted: boolean,
  topStories: Array<TopStories>,
  topStoryIds: Array<number>,
};

class App extends Component<{}, State> {
  state = {
    activePage: 1,
    isCompleted: false,
    topStories: [],
    topStoryIds: [],
  };

  componentDidMount() {
    axios
      .get(HN_TOP_STORIED_URL)
      .then((response) => {
        this.setState({ topStoryIds: response.data });
        this.state.topStoryIds.map((id, index) => this.getList(id, index));
      })
      .catch((error) => {
        console.log(error);
      });
  }

  getList = (id: number, index: number): void => {
    axios
      .get(`${HN_STORY_URL}/${id}.json`)
      .then(response => this.getItem(response.data, index))
      .catch((error) => {
        console.log(error);
      });
  };

  getItem = (item: TopStories, itemIndex: number): void => {
    itemArray[itemIndex] = item;
    this.setState({
      topStories: itemArray,
      isCompleted: true,
    });
  };

  getKidsList = (id: number, index: number, parentIndex: number): void => {
    axios
      .get(`${HN_STORY_URL}/${id}.json`)
      .then(response => this.getKidsItem(response.data, index, parentIndex))
      .catch((error) => {
        console.log(error);
      });
  };

  getKidsItem = (item: KidsItems, itemIndex: number, parentIndex: number): void => {
    textArray[itemIndex] = item;
    itemArray[parentIndex] = {
      ...itemArray[parentIndex],
      kidsItems: [...textArray],
    };
    this.setState({
      topStories: itemArray,
    });
  };

  handleGetKidsItems = (itemIndex: number): void => {
    if (itemArray[itemIndex].kids) {
      const itemArrayresults = itemArray[itemIndex].kids.map((id, index) =>
        this.getKidsList(id, index, itemIndex));
      Promise.all(itemArrayresults);
      textArray = [];
    }
  };

  handlePageChange = (pageNumber: number): void => {
    this.setState({ activePage: pageNumber });
  };

  render() {
    const { activePage, topStories, isCompleted } = this.state;
    const indexOfLastStories = activePage * 60;
    const indexOfFirstStories = indexOfLastStories - 60;
    const currentStories = topStories.slice(indexOfFirstStories, indexOfLastStories);
    return (
      <div>
        <header>
          <img src="https://news.ycombinator.com/y18.gif" alt="hackernews logo" />
          <h1>
            <a href="/">Hacker News</a>
          </h1>
        </header>

        {isCompleted ? (
          <div>
            <div className="App">
              {currentStories.map((story, index) => (
                <StoryItem
                  className="story_item"
                  key={story.id}
                  story={story}
                  handleGetKidsItems={() => this.handleGetKidsItems(index + indexOfFirstStories)}
                />
              ))}
            </div>
            <Pagination
              activePage={activePage}
              itemsCountPerPage={60}
              totalItemsCount={topStories.length}
              pageRangeDisplayed={5}
              onChange={this.handlePageChange}
              itemClass="pagination_item"
              linkClass="pagination_link"
              activeLinkClass="pagination_link--active"
            />
          </div>
        ) : (
          <div className="spinner">
            <div>
              <img
                // https://ph-files.imgix.net/b1d72c43-5634-4b47-b9d2-9b24c0d47df0?auto=format&auto=compress&codec=mozjpeg&cs=strip&fit=crop&w=50&h=50
                src="http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif"
                alt="orange hn spinner"
              />
            </div>
          </div>
        )}
      </div>
    );
  }
}

export default App;
