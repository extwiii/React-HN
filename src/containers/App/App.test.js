import React from 'react';
import { shallow } from 'enzyme';

import App from './index';

describe('App', () => {
  let mountedApp;
  const app = () => {
    if (!mountedApp) {
      mountedApp = shallow(<App />);
    }
    return mountedApp;
  };

  test.skip('renders correctly', () => {});

  test('renders correctly', () => {
    const divs = app().find('div');
    expect(divs.length).toBeGreaterThan(0);
    expect(divs.length).toBe(3);
    expect(app()).toMatchSnapshot();
  });

  beforeEach(() => {
    mountedApp = undefined;
  });
});
