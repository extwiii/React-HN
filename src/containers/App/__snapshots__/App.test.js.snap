// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header>
          <img
                    alt="hackernews logo"
                    src="https://news.ycombinator.com/y18.gif"
          />
          <h1>
                    <a
                              href="/"
                    >
                              Hacker News
                    </a>
          </h1>
</header>,
        <div
          className="spinner"
>
          <div>
                    <img
                              alt="orange hn spinner"
                              src="http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif"
                    />
          </div>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="hackernews logo"
              src="https://news.ycombinator.com/y18.gif"
/>,
            <h1>
              <a
                            href="/"
              >
                            Hacker News
              </a>
</h1>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "hackernews logo",
              "src": "https://news.ycombinator.com/y18.gif",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <a
                href="/"
>
                Hacker News
</a>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Hacker News",
                "href": "/",
              },
              "ref": null,
              "rendered": "Hacker News",
              "type": "a",
            },
            "type": "h1",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <img
                        alt="orange hn spinner"
                        src="http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif"
            />
</div>,
          "className": "spinner",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt="orange hn spinner"
              src="http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif"
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "orange hn spinner",
              "src": "http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header>
            <img
                        alt="hackernews logo"
                        src="https://news.ycombinator.com/y18.gif"
            />
            <h1>
                        <a
                                    href="/"
                        >
                                    Hacker News
                        </a>
            </h1>
</header>,
          <div
            className="spinner"
>
            <div>
                        <img
                                    alt="orange hn spinner"
                                    src="http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif"
                        />
            </div>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="hackernews logo"
                src="https://news.ycombinator.com/y18.gif"
/>,
              <h1>
                <a
                                href="/"
                >
                                Hacker News
                </a>
</h1>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "hackernews logo",
                "src": "https://news.ycombinator.com/y18.gif",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <a
                  href="/"
>
                  Hacker News
</a>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Hacker News",
                  "href": "/",
                },
                "ref": null,
                "rendered": "Hacker News",
                "type": "a",
              },
              "type": "h1",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              <img
                            alt="orange hn spinner"
                            src="http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif"
              />
</div>,
            "className": "spinner",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="orange hn spinner"
                src="http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif"
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "orange hn spinner",
                "src": "http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
